### PR TITLE
fix chain call issue in HTTP JSON input

### PR DESCRIPTION
**Pull request by kush-elastic**
## What does this PR do?

In HTTP JSON input of filebeat, chain calls were being executed and every response was collected first then they were published. So, it was using a lot of memory. Therefore filebeat was exiting without publishing any event.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in CHANGELOG.next.asciidoc or CHANGELOG-developer.next.asciidoc.

## Related issues
- Relates #34310 
- Relates https://github.com/elastic/sdh-beats/issues/2720

## Use cases
This PR contains changes to execute the publishing and collection of events concurrently. So, events will not be stored in memory while being collected and published later. Instead, events collection, processing, and publishing will be done simultaneously.

### CODE REVIEW

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:26]:
```
const lorem = "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum."
```

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:27]: Delete

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:36]:
```suggestion
	for i := 0; i < 2; i++ {
		for j := 0; j < 2500; j++ {
			expectedBulkData = append(expectedBulkData, fmt.Sprintf(`{ "text%d":%q }`, i+1, strings.Repeat(lorem, 17))
		}
	}
```

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:1421]:
```suggestion
				fmt.Fprintf(w, `{ "text1":%q }`+"\n", strings.Repeat(lorem, 17))
```

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:1425]:
```suggestion
				fmt.Fprintf(w, `{ "text2":%q }`+"\n", strings.Repeat(lorem, 17))
```

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:299]:
```suggestion
	defer func() {
		r.log.Infof("request finished: %d events published", n)
	}()
```
The deferred func is a closure, so it is already holding `n` as a reference to the enclosing scope's var `n`.

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:861]:
```suggestion
func closeChannels(chans ...chan *http.Response) {
	for _, c := range chans {
		close(c)
	}
}
```

Initially I had this as written below, but the non-`*http.Response` chans are never more than unity, so the genericisation is not required.

```
func closeChannels[E any](chans ...chan E) {
	for _, c := range chans {
		close(c)
	}
}
```
~~Then at the call sites you can `close(c)`, `closeChannels(c1, c2)` and  `closeChannels(c1, c2, c3)` (though there will need to be calls for each type of chan).~~

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:316]:
```suggestion
				close(publishedEventsCountChannel)
				close(events)
				closeChannels(responseChannel, intermediateResponseChannel, initialResponseChannel)
```
making use of the `closeChannels` below.

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:348]:
```suggestion
				closeChannels(responseChannel, intermediateResponseChannel)
```
and so on.

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:544]:
```suggestion
				outputChannel <- idsStruct{err: fmt.Errorf("error while reading response body: %w", err)}
```
Zero values do not need to be explicitly assigned in a decl — also below.

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:596]:
```suggestion
	outputChannel <- idsStruct{ids: ids}
```

File: x-pack/filebeat/input/httpjson/request.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/request.go:299]: Got it. I'll update it.

File: x-pack/filebeat/input/httpjson/request.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/request.go:861]: Okay, I'll update it accordingly.

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:45]: It would be good to have an explanation for why we are doing this. Particularly what the invariants are that `startSignalChannel` is ensuring are - a name change could help this, suggest `idsReady` or similar. I'm wondering also if you can't get away with a lock here (essentially what you are using is a waitgroup, but with only one member).
```diff
diff --git a/x-pack/filebeat/input/httpjson/request.go b/x-pack/filebeat/input/httpjson/request.go
index 49d08ca039..c7aad71c07 100644
--- a/x-pack/filebeat/input/httpjson/request.go
+++ b/x-pack/filebeat/input/httpjson/request.go
@@ -13,6 +13,7 @@ import (
        "net/http"
        "net/url"
        "strings"
+       "sync"
 
        "github.com/PaesslerAG/jsonpath"
 
@@ -41,7 +42,35 @@ type idsStruct struct {
 
 type processResponse struct {
        httpResponse       *http.Response
-       startSignalChannel chan struct{}
+       startSignalChannel *sync.Mutex
+}
+
+func block(resp *http.Response) processResponse {
+       var mu sync.Mutex
+       mu.Lock()
+       return processResponse{resp, &mu}
+}
+
+// On a pointer to get a side-effect. All the methods should agree, but this
+// is just demonstrational.
+func (p *processResponse) block() processResponse {
+       var mu sync.Mutex
+       mu.Lock()
+       p.startSignalChannel = &mu
+       return *p
+}
+
+func (p processResponse) done() {
+       if p.startSignalChannel != nil {
+               p.startSignalChannel.Unlock()
+       }
+}
+
+func (p processResponse) wait() {
+       if p.startSignalChannel != nil {
+               // This abuses sync.Mutex to behave as a waitgroup of one.
+               p.startSignalChannel.Lock()
+       }
 }
 
 func (c *httpClient) do(stdCtx context.Context, req *http.Request) (*http.Response, error) {
@@ -382,10 +411,7 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
                        }
 
                        // passing signal channel here to avoid closing of nil channel in getIdsFromResponses function
-                       intermediateResponseChannel <- processResponse{
-                               httpResponse:       httpResp,
-                               startSignalChannel: make(chan struct{}),
-                       }
+                       intermediateResponseChannel <- block(httpResp)
                        close(intermediateResponseChannel)
                        idsChannel := make(chan idsStruct, 1)
                        go r.getIdsFromResponses(intermediateResponseChannel, idsChannel, r.requestFactories[i+1].replace)
@@ -474,8 +500,7 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
                                resp := processResponse{httpResponse: httpResp}
                                // store data according to response type
                                if !(i == len(r.requestFactories)-1 && len(ids) != 0) {
-                                       resp.startSignalChannel = make(chan struct{})
-                                       intermediateResponseChannel <- resp
+                                       intermediateResponseChannel <- resp.block()
                                }
                                responseChannel <- resp
                        }
@@ -516,7 +541,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                        b, err = io.ReadAll(resp.httpResponse.Body)
                        if err != nil {
                                outputChannel <- idsStruct{err: fmt.Errorf("error while reading response body: %w", err)}
-                               close(resp.startSignalChannel)
+                               resp.done()
                                closeSignalChannels(intermediateResps)
                                return
                        }
@@ -525,7 +550,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                err = resp.httpResponse.Body.Close()
                if err != nil {
                        outputChannel <- idsStruct{err: fmt.Errorf("error closing response body: %w", err)}
-                       close(resp.startSignalChannel)
+                       resp.done()
                        closeSignalChannels(intermediateResps)
                        return
                }
@@ -534,14 +559,14 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                var v interface{}
                if err := json.Unmarshal(b, &v); err != nil {
                        outputChannel <- idsStruct{err: fmt.Errorf("cannot unmarshal data: %w", err)}
-                       close(resp.startSignalChannel)
+                       resp.done()
                        closeSignalChannels(intermediateResps)
                        return
                }
                values, err := jsonpath.Get(replace, v)
                if err != nil {
                        outputChannel <- idsStruct{err: fmt.Errorf("error while getting keys: %w", err)}
-                       close(resp.startSignalChannel)
+                       resp.done()
                        closeSignalChannels(intermediateResps)
                        return
                }
@@ -554,7 +579,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                                        ids = append(ids, fmt.Sprintf("%v", v))
                                default:
                                        r.log.Errorf("events must a number or string, but got %T: skipping", v)
-                                       close(resp.startSignalChannel)
+                                       resp.done()
                                        continue
                                }
                        }
@@ -563,7 +588,7 @@ func (r *requester) getIdsFromResponses(intermediateResps <-chan processResponse
                default:
                        r.log.Errorf("cannot collect IDs from type '%T' : '%v'", values, values)
                }
-               close(resp.startSignalChannel)
+               resp.done()
        }
        outputChannel <- idsStruct{ids: ids}
 }
@@ -661,10 +686,7 @@ func (r *requester) processChainPaginationEvents(stdCtx context.Context, trCtx *
        idsChannel := make(chan idsStruct, 1)
 
        // passing signal channel here to avoid closing of nil channel in getIdsFromResponses function
-       intermediateResponseChannel <- processResponse{
-               httpResponse:       response,
-               startSignalChannel: make(chan struct{}),
-       }
+       intermediateResponseChannel <- block(response)
        close(intermediateResponseChannel)
 
        go r.getIdsFromResponses(intermediateResponseChannel, idsChannel, r.requestFactories[chainIndex].replace)
@@ -742,8 +764,7 @@ func (r *requester) processChainPaginationEvents(stdCtx context.Context, trCtx *
 
                        resp := processResponse{httpResponse: httpResp}
                        if !(i == len(r.requestFactories)-1 && len(ids) != 0) {
-                               resp.startSignalChannel = make(chan struct{})
-                               intermediateResponseChannel <- resp
+                               intermediateResponseChannel <- resp.block()
                        }
                        processingChannel <- resp
                }
@@ -818,6 +839,6 @@ func closeChannels(channels ...chan processResponse) {
 
 func closeSignalChannels(channel <-chan processResponse) {
        for c := range channel {
-               close(c.startSignalChannel)
+               c.done()
        }
 }
diff --git a/x-pack/filebeat/input/httpjson/response.go b/x-pack/filebeat/input/httpjson/response.go
index 7b800a0162..84d8c86813 100644
--- a/x-pack/filebeat/input/httpjson/response.go
+++ b/x-pack/filebeat/input/httpjson/response.go
@@ -136,9 +136,7 @@ func (rp *responseProcessor) startProcessing(stdCtx context.Context, trCtx *tran
 
        for resp := range resps {
                // in case of intermediate responses wait for IDs to be collected
-               if resp.startSignalChannel != nil {
-                       <-resp.startSignalChannel
-               }
+               resp.wait()
 
                iter := rp.pagination.newPageIterator(stdCtx, trCtx, resp.httpResponse)
                for {
```

**Review by efd6 — COMMENTED** [review]
The invariant docs need to be in place, the diff is suggestion to consider.

File: x-pack/filebeat/input/httpjson/request.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/request.go:45]: While testing we found that the intermediate response was getting published instead of final response while using double chain calls.

**Cause**: 
HTTP response body should be read first by getIdsFromResponses function and then by startProcessing function. So, in concurrent execution of both of these function the response body may be read by startProcessing function first.

**Solution**:
In startProcessing function wait for getIdsFromResponses function to finish collecting IDs from the response.

I'll update the PR with these changes.

Comment (efd6) [reply x-pack/filebeat/input/httpjson/request.go:45]: The code above is not production-ready, so if you use it, please clean it up.

File: x-pack/filebeat/input/httpjson/request.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/request.go:71]:
```suggestion
type processResponse struct {
	httpResponse *http.Response
	
	// idsReady ensures that a response will not be processed
	// until double chain calls responses have been finalized.
	// While idsReady is a sync.Mutex it is being used here as
	// a waitgroup with a single member. This means that the
	// wait method may only be called once, otherwise a deadlock
	// will result. If multiple wait calls are required in the future,
	// the type should be replaced with a sync.WaitGroup.
	idsReady *sync.Mutex
}

// block returns a processResponse that will block until response completion.
func block(resp *http.Response) processResponse {
	var mu sync.Mutex
	mu.Lock()
	return processResponse{resp, &mu}
}

// block upgrades the receiver to a blocking state. It may only be called once
// in the lifetime of the processResponse. The upgraded value is returned.
func (p *processResponse) block() processResponse {
	var mu sync.Mutex
	mu.Lock()
	p.idsReady = &mu
	return *p
}

// done releases the receiver block. It may only be called once in the lifetime
// of the processResponse.
func (p *processResponse) done() {
	if p.idsReady != nil {
		p.idsReady.Unlock()
	}
}

// wait will block until the receiver's block has been released. In the current
// implementation it may only be called once in the lifetime of the processResponse.
func (p *processResponse) wait() {
	if p.idsReady != nil {
		p.idsReady.Lock()
	}
}
```

File: x-pack/filebeat/input/httpjson/request.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/request.go]: Thanks for your suggestions. I have updated the waiting logic to use mutex instead of channel.

**Review by efd6 — APPROVED** [review]
Thanks

**Comment by ShourieG** [comment]
@kush-elastic In order to start on the PR https://github.com/elastic/beats/issues/34839, we need this merged as the fix will non trivial changes in various overlapping code blocks changed in this PR.

**Comment by kush-elastic** [comment]
> @kush-elastic In order to start on the PR #34839, we need this merged as the fix will non trivial changes in various overlapping code blocks changed in this PR.

@lalit-satapathy, let us know if we can have your review over this PR. so we can merge and unblock @ShourieG.

**Comment by ShourieG** [comment]
@kush-elastic can we merge this  ?

**Comment by ebeahan** [comment]
@kush-elastic are you awaiting the next round of reviews here, or anything else we can do to help move this PR along?

**Comment by kush-elastic** [comment]
> @kush-elastic can you resolve the conflicts so we can we go ahead and merge this ?

I have resolved the merge conflicts. let's wait for CI to pass.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:35]: I would prefer that this change (placing `testCases` into a `var ( )` block did not happen. It brings unnecessary diff noise at no benefit.

Also, let's make `expectedBulkData` a local var in the test function body.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:30]: Let's move this to below the test; it's not important what it is, so it doesn't need to be so promenant in the file.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:1373]:
```suggestion
	bulkData := make([]string, 5000)
	text := strings.Repeat(dummyText, 17)
	for i := range bulkData {
		bulkData[i] = fmt.Sprintf(`{ "text%d":%q }`, (i/2500)+1, text)
	}
```
Though, I'd put this in a helper and have it called below where it's used since we don't need it more than once.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/input_test.go]: I have moved it below test cases.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/input_test.go]: Got it. Its done.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/input_test.go]: For now, I am keeping it there only. if we needed it anywhere else we will just move it to the helper function.

Comment (efd6) [reply x-pack/filebeat/input/httpjson/input_test.go]: I'd prefer it be moved. Suggest something like this
```diff
diff --git a/x-pack/filebeat/input/httpjson/input_test.go b/x-pack/filebeat/input/httpjson/input_test.go
index 2626ed1050..e8ca87c08e 100644
--- a/x-pack/filebeat/input/httpjson/input_test.go
+++ b/x-pack/filebeat/input/httpjson/input_test.go
@@ -36,6 +36,7 @@ var testCases = []struct {
        expectedFile string

        skipReason string
+       isLongTest bool
 }{
        {
                name:        "simple_GET_request",
@@ -813,6 +814,7 @@ var testCases = []struct {
        },
        {
                name:        "pagination_when_used_with_chaining_with_bulk_data",
+               isLongTest:  true,
                setupServer: newChainPaginationLoadTestServer(httptest.NewServer),
                baseConfig: map[string]interface{}{
                        "interval":       1,
@@ -836,6 +838,14 @@ var testCases = []struct {
                        },
                },
                handler: defaultHandler(http.MethodGet, "", ""),
+               expected: func() []string {
+                       bulkData := make([]string, 5000)
+                       text := strings.Repeat(dummyText, 17)
+                       for i := range bulkData {
+                               bulkData[i] = fmt.Sprintf(`{ "text%d":%q }`, (i/2500)+1, text)
+                       }
+                       return bulkData
+               }(),
        },
        {
                name: "replace_with_clause_and_first_response_object",
@@ -1362,11 +1372,6 @@ var testCases = []struct {
 const dummyText = "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum."

 func TestInput(t *testing.T) {
-       bulkData := make([]string, 5000)
-       text := strings.Repeat(dummyText, 17)
-       for i := range bulkData {
-               bulkData[i] = fmt.Sprintf(`{ "text%d":%q }`, (i/2500)+1, text)
-       }
        for _, test := range testCases {
                t.Run(test.name, func(t *testing.T) {
                        if test.skipReason != "" {
@@ -1401,12 +1406,9 @@ func TestInput(t *testing.T) {
                        g.Go(func() error {
                                return input.Run(ctx, chanClient)
                        })
-                       var timeoutDuration time.Duration
-                       if test.name == "pagination_when_used_with_chaining_with_bulk_data" {
+                       timeoutDuration := 10 * time.Second
+                       if test.isLongTest {
                                timeoutDuration = 2 * time.Minute
-                               test.expected = bulkData
-                       } else {
-                               timeoutDuration = 10 * time.Second
                        }
                        timeout := time.NewTimer(timeoutDuration)
                        t.Cleanup(func() { _ = timeout.Stop() })
```

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:1205]: Please revert this whitespace change.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:1331]: Please revert this whitespace change.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/input_test.go]: Okay, Got it. I will move it as you suggested.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/input_test.go:1205]: On it.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/input_test.go:1331]: On it.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (efd6) [inline x-pack/filebeat/input/httpjson/input_test.go:1119]: Please reverse this whitespace change as well. also at L1198.

**Review by efd6 — COMMENTED** [review]
LGTM after final nits.

File: x-pack/filebeat/input/httpjson/input_test.go
Comment (kush-elastic) [reply x-pack/filebeat/input/httpjson/input_test.go]: Done.

**Review by efd6 — APPROVED** [review]

**Review by ShourieG — CHANGES_REQUESTED** [review]
@kush-elastic after having gone through the code and having some internal discussions, this is certainly not a trivial change as it changes the fundamentals of chaining and how it functions. Since httpjson is used in most of our integrations we need to be extra careful in implementing major overhauls, so I kindly ask that the following be done :- 

1) Keep the old logic as a fallback that can be easily toggled on/off if the need arises, this is extremely important as it will help us avoid a lot of hassle and user outrage if something breaks-down at scale. This fallback can be removed later after the current code is stable for a while. This change is no-trivial and will possibly require a re-work of the logic. The fallback toggle could be done as part of the config as then it would be very easy to toggle it on/off. 

2) Maintain an official documentation detailing the changes to functionality in the current flow and add flow diagrams if necessary.

3) Add more meaningful tests that uses data at scale/large amounts along with tests for the fallback mechanism also.

With these changes, I feel this PR would be ready to be merged.

**Comment by jamiehynds** [comment]
@kush-elastic is this non-trivial change purely related to Salesforce or are there other integrations that require this new approach to chained calls? Since this PR was originally raised, we've added a new [CEL input](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-cel.html), which may be a better fit for this use case, and avoid making fundamental changes to httpjson just for the Salesforce integration. 

While we don't have a perfect approach to switching inputs in an existing integration, we do have some options which we can walk through. We are currently switching inputs on our Office 365 integration. 

cc @SubhrataK

**Comment by narph** [comment]
@kush-elastic , @SubhrataK , just a heads up here, we have planned efforts to improve and simplify the httpjson input https://github.com/elastic/beats/issues/35432. After some discussion we've decided to hold off on the changes from this PR until the refactoring is complete and reassess the functionality after.

**Comment by ShourieG** [comment]
Closing this issue as this was handled in the following meta issue : https://github.com/elastic/beats/issues/35432

**Pull request closed without merging**
